### ISSUE TITLE
Use FIPS-approved encryption

### ISSUE THREAD

**Issue by cecilia-donnelly** [SolutionGuidance/psm#465]
"The PSM shall use only FIPS Pub 140-2-approved (or higher) encryption algorithms."  and "The Contractor's data encryption solution shall meet Federal Information Processing Standard (FIPS) 140-2 and at a minimum use AES 128 encryption."

The first step here is to get the list of FIPS-approved algorithms (I think @jasonaowen may know about this?).  See also #104.

**Comment by cecilia-donnelly** [comment]
@jasonaowen tells us that he's worked with this before and we may well be able to set [`bouncycastle`](https://www.bouncycastle.org/) to do the right thing and then make any necessary code changes.  Finally, we'll validate that it works as intended.

**Comment by jvasile** [comment]
I'm ok letting bouncycastle DTRT, but I do think we should list all our encryption uses in this issue and check them all explicitly.

This  doc might help: https://csrc.nist.gov/csrc/media/publications/fips/140/2/final/documents/fips1402annexa.pdf

> The purpose of this document, and of Annexes C and D, is to provide a list of the approved security functions applicable to FIPS 140-2. Annex C lists the approved Random Bit Generators, while Annex D shows the approved Key Establishment Methods. The remaining approved security functions are listed in this Annex. The Annexes also provide the links to the descriptions of the allowed algorithms.

@jasonaowen Can you tell me where we are using encryption in the app? 

Two quick things come to mind on action items:
* We should tune the SSL layer so the server only supports certain algorithms.  
* We're hashing passwords, yes? SHA3-256 should be our target there.

What else is encrypted?

**Comment by jvasile** [comment]
We *might* want to mention FIPS in the docs so that future integration efforts keep it in mind.  We should make sure our scripts DTRT for SSL to any proxy servers.

**Comment by jvasile** [comment]
Actually, before we do all this work, maybe this is the moment to move to SAML and offload some of this to our SSO provider.  Otherwise we do this analysis once and again when we mote to SAML.  @jasonaowen, I'd appreciate your thoughts on tradeoffs here.

**Comment by jvasile** [comment]
After consultation with @jasonaowen we're going to (a) make sure we know what security code we have and (b) make sure that code exclusively uses bouncycastle. That should get us what we need. We'll put off auth delegation for another day.

**Comment by cecilia-donnelly** [comment]
Some ideas of what data should be stored encrypted:
- [ ] SSN
- [ ] EIN
- [ ] State tax ID
- [ ] EFT vendor number (though see #411)
- [ ] State Medicaid ID (maybe?)

A+ on that plan, @jvasile.

**Comment by jvasile** [comment]
We have encryption in the login process, SSL that wraps all traffic on 443 (but do we need it for the ETL stuff?), and hashed passwords.  Is that the full list?

**Comment by jasonaowen** [comment]
> We have encryption in the login process, SSL that wraps all traffic on 443 (but do we need it for the ETL stuff?), and hashed passwords. Is that the full list?

Yes, I believe so. The login process we're familiar with is the password-in-the-database method, but there's also some LDAP code floating around that we haven't spent any time on (see also the discussion in #34 ). We may need to use HTTPS for some of the external data services, depending on where they live, but right now we only have the LEIE which we serve over HTTP.

For the passwords we store in the database, we should use [PBKDF2](https://en.wikipedia.org/wiki/PBKDF2): if used with a FIPS-compliant pseudorandom function (such as SHA1), then PBKDF2 is FIPS-compliant. And, indeed, [Spring Security has support for PBKDF2](https://docs.spring.io/spring-security/site/docs/4.2.3.RELEASE/reference/html/crypto.html).

Currently, we load the "SHA" algorithm in a few places ([1](https://github.com/OpenTechStrategies/psm/blob/bcab22c4ec8ccc5f7b2c7b2a9519a5c1b09d032f/psm-app/cms-business-process/src/main/java/gov/medicaid/services/impl/DBIdentityProviderDAOBean.java#L123), [2](https://github.com/OpenTechStrategies/psm/blob/bcab22c4ec8ccc5f7b2c7b2a9519a5c1b09d032f/psm-app/services/src/main/java/gov/medicaid/dao/impl/LDAPIdentityProviderDAOBean.java#L422), [3](https://github.com/OpenTechStrategies/psm/blob/bcab22c4ec8ccc5f7b2c7b2a9519a5c1b09d032f/psm-app/services/src/main/java/gov/medicaid/services/util/Util.java#L218)), which winds up being SHA1. You can verify this by comparing a known password (`p1`) with one in the database:

```ShellSession
$ echo "SELECT password FROM cms_authentication WHERE username = 'p1';" \
  | psql -h localhost -U psm psm
             password              
-----------------------------------
 {SHA}t49XZhHsBvlq88plTCIXKl10bEA=
(1 row)
$ echo -n t49XZhHsBvlq88plTCIXKl10bEA= | base64 -d | xxd
00000000: b78f 5766 11ec 06f9 6af3 ca65 4c22 172a  ..Wf....j..eL".*
00000010: 5d74 6c40                                ]tl@
$ echo -n p1 | sha1sum
b78f576611ec06f96af3ca654c22172a5d746c40  -
```

The SHA series of hashing algorithms are not appropriate for passwords. From [About Secure Password Hashing, on Security.SE](http://security.blogoverflow.com/2013/09/about-secure-password-hashing/):

> A password hashing algorithm should be slow to prevent bruteforce attacks, [preferably] it should have features which actually decrease the feasibility of a distributed brute force attack on the hashes. This immediately throws out [SHA-1 and SHA-3].

After chatting briefly about this with @cecilia-donnelly , switching our password hashing to PBKDF2 will be my next priority.

I also looked into how to configure WildFly to be FIPS-compliant. It looks to be a combination of operating system configuration and selecting FIPS-compliant algorithms. See:
- [Configure a FIPS 140-2 Compliant Java Provider on RedHat/CentOS/Fedora](http://blog.johnruiz.com/2013/12/configure-fips-140-2-compliant-java.html)
- If we need FIPS certification (rather than merely compliance), I think we would need to use the RedHat version of WildFly, [RedHat JBoss Enterprise Application Platform](https://access.redhat.com/products/red-hat-jboss-enterprise-application-platform/#knowledge)
- [RedHat documentation on configuring JBoss EAP 7](https://access.redhat.com/documentation/en-us/red_hat_jboss_enterprise_application_platform/7.1.beta/html/how_to_configure_server_security/securing_the_server_and_its_interfaces#configure_ssl_fips_rhel7)

**Comment by cecilia-donnelly** [comment]
@jasonaowen, what's the status of this?  I know we still need to fix #34, but can we close this FIPS issue?

**Comment by cecilia-donnelly** [comment]
@jasonaowen confirms that we can close this.
